index.d.ts

```diff
@@ -303,24 +334,31 @@ declare class Emittery {
 		});
 	```
 	*/
-	anyEvent(): AsyncIterableIterator<unknown>
+	anyEvent(): AsyncIterableIterator<
+		[keyof EventData, EventData[keyof EventData]]
+	>;
 
 	/**
 	Remove an `onAny` subscription.
 	*/
-	offAny(listener: (eventName: EventName, eventData?: unknown) => void): void;
+	offAny(
+		listener: <Event extends keyof EventData>(
+			eventName: Event,
+			eventData?: EventData[Event]
+		) => void
+	): void;
 
 	/**
 	Clear all event listeners on the instance.
 
 	If `eventName` is given, only the listeners for that event are cleared.
 	*/
-	clearListeners(eventName?: EventNames): void;
+	clearListeners(eventName?: keyof EventData): void;
 
 	/**
 	The number of listeners for the `eventName` or all events if not specified.
 	*/
-	listenerCount(eventName?: EventNames): number;
+	listenerCount(eventName?: keyof EventData): number;
```

Comment:
Maybe make a type alias for `type EventNames = keyof EventData`?

Reply:
I opted not to do that because to create type aliases from operators on the generics within class definitions like this, you have to use more generic types with defaults. To the best of my knowledge, you can't just do:

```typescript
class Foo<T> {
  type U = keyof T;
  // ...
}
```

you have to do 

```typescript
class Foo<T, U = keyof T> {
  // ...
}
```

For types that are meant to be consumed by end users, which `Emittery` definitely is, I always find this confusing. The type you are going to use has more type parameters than you are supposed to pass, and without looking at the code or knowing that the code is using this trick, it seems unclear. So, I'd say its better to be verbose in the class body and keep the top level user facing type interface clean, but happy to change if you feel strongly.

Reply:
Just realized I did this already for the `DatalessEvents` type so I am being kind of inconsistent, oops. I will move this up in the same way.

Reply:
AH I remember why this wasn't possible now -- fancy typesystem reasons. Because of the strange no-type-aliases-only-derived-generic-type-defaults thing above, the generic type will usually be the `keyof EventNames` type, but could be instantiated with a different type if the user passed it in. For that reason, TypeScript can't for sure, always access `EventNames[Name]`, because `Name` only defaults to `keyof EventNames`, but could be something else. So, it doesn't type check, so I can't really do it. Sorry for all the confusion!